### ISSUE TITLE
Remove Starcounter bin dependency

### ISSUE THREAD

**Issue by miyconst** [Starcounter/StarDump#57]
Currently StarDump uses `star.exe` to read current Starcounter version, and Administrator REST API to read database settings.

This should be removed in order to make it compatible with Linux environment and NuGet package installation.

@ubbeK please specify which values we need in StarDump.

**Comment by ubbeK** [comment]
It is only `star.exe -v` that is currently needed for the tool itself to work properly. 

But for test purposes it would be nice to have:
Extracting all database names, `staradmin.exe list db`
create new database `staradmin.exe --database=XYZ new db DefaultUserHttpPort=1234`
delete database, `staradmin.exe --database=XYZ delete --force db`

cc @linkdata

**Comment by miyconst** [comment]
@linkdata if there is no options for `Starcounter.Core` to retrieve Starcounter version, then we need to involve @bigwad to implement something for us.

Yes, I remember the conversation, when you and @bigwad said that we don't have version for the database engine, only Starcounter installation.

Probably, it's time to fix it.

**Comment by ubbeK** [comment]
In current build environment, the [BUILD_NUMBER](https://github.com/Starcounter/CI.BuildSystem/blob/develop/src/Scripts/teamcity.py#L88-L92) is extracted and set as an environment variable before compiling `BlueCmake.sln` and `Level1.sln`

**Comment by ubbeK** [comment]
It is inserted into `VersionInfo.xml` through [GenerateInstaller.exe](https://github.com/Starcounter/CI.BuildSystem/blob/develop/src/GenerateInstaller/GenerateInstaller.cs#L322-L383) which is executed in the end of the CI chain. There is no technical reason why this even cant be inserted into BlueCMake before compiling it.

**Comment by linkdata** [comment]
Those tools are part of `level1`, and as such Windows-only.

Version info should be available for components both in a file and using an API on the component. For things that are distributed with NuGet, that information can be gotten using NuGet tooling and APIs.

Listing databases is hard simply because the configuration system as a whole is organically evolved rather than designed. So that needs to be designed. But to preserve compatibility with existing product, some complexity must remain. At it's lowest level, a database is just a directory with files in it, and what we need is a good way to detect a database directory (currently it's looking for files with certain extensions in it) and some standard places to look for those directories (currently that means using `StarcounterBin` and parsing a chain of XML files).

Deleting databases is just a matter of deleting files, as long as noone is using them. The user of those files is `Bluestar`. So database deletion should probably be an API and a utility in `level0`.

So to summarize my opinions:

* Versioning can be resolved by moving to NuGet.
* Configuration needs to be designed, and have a standard package capable of deciphering legacy Windows installations.
* Low-level database management CLI tools should be written i C/C++ in `level0`, not C# `level1`.

**Comment by ubbeK** [comment]
If we skip the add/remove database functionality for a while and concentrate on the version only. We currently save the Installer Starcounter version in a StarDump file. StarDump itself only have reference to Starcounter.Core nuget package. 

> Version info should be available for components both in a file and using an API on the component. For things that are distributed with NuGet, that information can be gotten using NuGet tooling and APIs.

So by this reasoning we should not use the Installer version but rather the version of Starcounter.Core nuget package. But Starcounter.Core would still need to supply some API to show what versions of its dependencies are used, at least the BlueStar nuget dependency or what it is called.

**Comment by miyconst** [comment]
Yes, we have to save version of database engine, not the `Starcounter.Core` nuget package. This information is absolutely required for debugging and compatibility purposes.

**Comment by ubbeK** [comment]
> Yes, we have to save version of database engine, not the Starcounter.Core nuget package. This information is absolutely required for debugging and compatibility purposes.

I would say that for 2.3.1.x we need the Installer version in the dumps, but when 2.4 is out, and the concept of Installer is removed, then we need the versions of all the packages.

**Comment by miyconst** [comment]
Basically, we need any number which would let us uniquely identify Starcounter Database Engine and Code Host.

**Comment by ubbeK** [comment]
@linkdata But I dont want StarDump to be aware of Starcounter.Core:s dependencies. What if Starcounter.Code starts to reference another Starcounter dependency or if BlueStar is split into two separate packages, then this change would have to be extracted in StarDump aswell. I would prefer it Starcounter.Core had an API for this, a dictionary with assembly as key and version as value.

**Comment by linkdata** [comment]
> Basically, we need any number which would let us uniquely identify Starcounter Database Engine and Code Host.

> But I dont want StarDump to be aware of Starcounter.Core:s dependencies.

Aren't those statement mutually exclusive?

Either you just need the SC.Core version and don't care about it's dependencies, or you need the Bluestar version too, and then you do care.

**Comment by linkdata** [comment]
> I would prefer it Starcounter.Core had an API for this, a dictionary with assembly as key and version as value.

What is it you need the version information for? If it's for dumping info to the user, I think iterating over all loaded assemblies with `Starcounter` in them would probably be best.

We can certainly write an API that abstracts away the differences between .NET and .NET Core and just returns you a list of loaded assemblies with paths and versions.

However, that's a really generic piece of code and would be better off in it's own utility assembly rather than packages with `Starcounter.Core`.

**Comment by ubbeK** [comment]
> Aren't those statement mutually exclusive?
> 
> Either you just need the SC.Core version and don't care about it's dependencies, or you need the Bluestar version too, and then you do care.

This has been taken out of context, please read next line aswell. To claify: StarDump is dependent on all Starcounter.Core:s dependencies, but I don't want to write code which extracts Starcounter.Core:s dependencies in StarDump.

**Comment by ubbeK** [comment]
> What is it you need the version information for? If it's for dumping info to the user, I think iterating over all loaded assemblies with Starcounter in them would probably be best.

We need to know the state of Starcounter database when dumping the it for debugging and compatibility purposes when reloading a dump again.

**Comment by miyconst** [comment]
We also have to keep in mind, that we aim to support multiple Starcounter setups on same machine.

**Comment by linkdata** [comment]
Ok, then I suggest we do what I wrote:

* Write an API that abstracts away the differences between .NET and .NET Core and just returns a list of loaded assembly names with paths and versions, all as strings.
* Place this API in it's own project and publish it as a NuGet package targeting multiple frameworks, at a minimum .NET 4.5.1, .NET 4.6.1, .NET Core 1.1 and .NET Core 2.0.

I think this might be a nice bite-sized project for someone learning C#.

**Comment by ubbeK** [comment]
Sounds good. 

So this API would be called in StarDump **after** it has connected to the db host through `Starcounter.Core.Hosting.AppHostBuilder`? 

> just returns a list of loaded assembly names with paths and versions, all as strings.

Would this be all assemblies, not only Starcounter created assemblies?

**Comment by linkdata** [comment]
The API would not depend on anything Starcounter related, but it can only be relied to enumerate things that are loaded. Maybe call it after building the host object?

I would make it so that it took a `StartsWith()` filter argument. Passing `null` would return everything.

```cs
public class AssemblyInfo {
  string Name;
  string Path;
  string Version;
  // other stuff..?
};
public IEnumerable<AssemblyInfo> GetLoadedAssemblies(string startsWithFilter = null);
```

**Comment by ubbeK** [comment]
Ok, thanks!

the startsWithFilter would preferably be a string array though.

**Comment by linkdata** [comment]
> the startsWithFilter would preferably be a string array though.

I think if you want more advanced filtering you're better off getting everything and filtering yourself. Or supply a delegate that accepts or rejects if you want to get fancy.

**Comment by linkdata** [comment]
```cs
public delegate bool AssemblyFilterDelegate(AssemblyInfo a);
public IEnumerable<AssemblyInfo> GetLoadedAssemblies(AssemblyFilterDelegate filter = null);
```
